x/btcstaking/types/incentive.go

```diff
@@ -36,21 +36,39 @@ func (dc *VotingPowerDistCache) FindNewActiveFinalityProviders(prevDc *VotingPow
 }
 
 // ApplyActiveFinalityProviders sorts all finality providers, counts the total voting
-// power of top N finality providers, and records them in cache
+// power of top N finality providers, excluding those who don't have timestamped pub rand
+// and records them in cache
 func (dc *VotingPowerDistCache) ApplyActiveFinalityProviders(maxActiveFPs uint32) {
-	// reset total voting power
-	dc.TotalVotingPower = 0
-	// sort finality providers
-	SortFinalityProviders(dc.FinalityProviders)
-	// calculate voting power of top N finality providers
-	numActiveFPs := dc.GetNumActiveFPs(maxActiveFPs)
+	// sort finality providers with timestamping considered
+	SortFinalityProvidersWithTimestamping(dc.FinalityProviders)
+
+	numActiveFPs := uint32(0)
+
+	for _, fp := range dc.FinalityProviders {
+		if numActiveFPs == maxActiveFPs {
+			break
+		}
+		if fp.TotalVotingPower == 0 {
+			break
+		}
```

Comment:
I remember `dc` will exclude FPs with zero total voting power. Is this true?

Reply:
Yes, but we don't have any rules to enforce this. This is a nice to have